Wrappers/Python/test/test_algorithms.py

```diff
@@ -990,4 +1019,98 @@ def callback (iteration, objective, solution):
 
         algo = PrintAlgo(update_objective_interval = 4, max_iteration = 1000)
 
-        algo.run(20, verbose=2, print_interval=2)
\ No newline at end of file
+        algo.run(20, verbose=2, print_interval=2)
+
+class TestADMM(unittest.TestCase):
+    def setUp(self):
+        ig = ImageGeometry(2,3,2)
+        data = ig.allocate(1, dtype=np.float32)
+        noisy_data = data+1
+        
+        # TV regularisation parameter
+        self.alpha = 1
+
+    
+
+        self.fidelities = [ 0.5 * L2NormSquared(b=noisy_data), L1Norm(b=noisy_data), 
+            KullbackLeibler(b=noisy_data, use_numba=False)]
+
+        # Setup and run the PDHG algorithm
+        G = ZeroFunction()
+        K = BlockOperator(GradientOperator(ig), IdentityOperator(ig))
+        # Compute operator Norm
+        normK = K.norm()
+
+        # Primal & dual stepsizes
+        self.sigma = 1./normK
+        self.tau = 1./normK
+        self.G = G
+        self.K = K
+
+    def test_ADMM_L2(self):
+        self.do_test_with_fidelity(self.fidelities[0])
+    def test_ADMM_L1(self):
+        self.do_test_with_fidelity(self.fidelities[1])
+    def test_ADMM_KL(self):
+        self.do_test_with_fidelity(self.fidelities[2])
+    def do_test_with_fidelity(self, fidelity):
+        alpha = self.alpha
+        F = BlockFunction(alpha * MixedL21Norm(),fidelity)
+        K = self.K
+        G = self.G 
+
+        admm = LADMM(f=G, g=F, operator=K, tau=self.tau, sigma=self.sigma,
+                    max_iteration = 100, update_objective_interval = 10)
+        admm.run(1, verbose=0)
+
+        admm_noaxpby = LADMM(f=G, g=F, operator=K, tau=self.tau, sigma=self.sigma,
+                    max_iteration = 100, update_objective_interval = 10, use_axpby=False)
+        admm_noaxpby.run(1, verbose=0)
+        
+        np.testing.assert_array_almost_equal(admm.solution.as_array(), admm_noaxpby.solution.as_array())
+
+    def test_compare_with_PDHG(self):
+        # Load an image from the CIL gallery. 
+        data = dataexample.SHAPES.get()
+        ig = data.geometry    
+        # Add gaussian noise
+        noisy_data = applynoise.gaussian(data, seed = 10, var = 0.005)
+
+        # TV regularisation parameter
+        alpha = 1
+
+        # fidelity = 0.5 * L2NormSquared(b=noisy_data)
+        # fidelity = L1Norm(b=noisy_data)
+        fidelity = KullbackLeibler(b=noisy_data, use_numba=False)
+
+        # Setup and run the PDHG algorithm
+        F = BlockFunction(alpha * MixedL21Norm(),fidelity)
```

Comment:
If we use `BlockFunction` and `BlockOperator` it will still work without the [tmp](https://github.com/vais-ral/CCPi-Framework/pull/825/files#diff-69f9f45b780b5c2fca8b56ef5fb5455c9f5949ae41a6e56e0b39b02bff2a579eR119)

Instead, use `F=alpha*MixedL21Norm()` ,`G = fidelity`  , `K = GradientOperator(ig)`

Reply:
What's the difference between `out[i]` and `x.get_item(i)`???

Reply:
There's no difference[ `__getitem__`](https://github.com/vais-ral/CCPi-Framework/blob/0a3ad90eaa0e8d1ef1ffedc809b79051cccf6c55/Wrappers/Python/cil/framework/BlockDataContainer.py#L129) is an alias of `get_item`

Reply:
I still cannot explain why the solutions are exactly when
1. we use the tmp trick 
2. without the tmp trick
under the BlockOperator/BlockFunction setup.

If we do not use BlockOperator/BlockFunction there is a difference

Reply:
I changed the formulation of the optimisation problem as @epapoutsellis suggested.

Unittests seem to work. Waiting for Jenkins.